v2/piv/piv.go

```diff
@@ -457,32 +497,54 @@ func ykAuthenticate(tx *scTx, key [24]byte, rand io.Reader) error {
 // are triple-des keys, however padding isn't verified. To generate a new key,
 // generate 24 random bytes.
 //
+// Note: Yubikeys also support aes128, aes192, and aes256 management keys,
+// which are 16, 24, and 32 bytes, respectively.
+//
 //	var newKey [24]byte
 //	if _, err := io.ReadFull(rand.Reader, newKey[:]); err != nil {
 //		// ...
 //	}
-//	if err := yk.SetManagementKey(piv.DefaultManagementKey, newKey); err != nil {
+//	if err := yk.SetManagementKey(piv.DefaultManagementKey, newKey[:]); err != nil {
 //		// ...
 //	}
-func (yk *YubiKey) SetManagementKey(oldKey, newKey [24]byte) error {
-	if err := ykAuthenticate(yk.tx, oldKey, yk.rand); err != nil {
+func (yk *YubiKey) SetManagementKey(oldKey, newKey []byte) error {
+	if err := ykAuthenticate(yk.tx, oldKey, yk.rand, yk.version); err != nil {
 		return fmt.Errorf("authenticating with old key: %w", err)
 	}
-	if err := ykSetManagementKey(yk.tx, newKey, false); err != nil {
+	if err := ykSetManagementKey(yk.tx, newKey, false, yk.version); err != nil {
 		return err
 	}
 	return nil
 }
 
 // ykSetManagementKey updates the management key to a new key. This requires
 // authenticating with the existing management key.
-func ykSetManagementKey(tx *scTx, key [24]byte, touch bool) error {
+func ykSetManagementKey(tx *scTx, key []byte, touch bool, version *version) error {
+	var managementKeyType byte
+	if supportsVersion(version, 5, 4, 0) {
```

Comment:
As described in https://github.com/go-piv/piv-go/issues/146#issuecomment-2176449257, AES support was only added in 5.4.2, not 5.4.0.

Reply:
Yubico claims keys 5.4+ support AES keys. If it can be confirmed a 5.4.0 or 5.4.1 key doesn't support them, I'd say we should fix that. Otherwise, I went with their documentation here (perhaps there is a different page of their docs that conflicts this?): https://developers.yubico.com/PIV/Introduction/YubiKey_and_PIV.html

<img width="844" alt="Screenshot 2024-07-10 at 07 27 56" src="https://github.com/go-piv/piv-go/assets/673216/34937cd2-fe50-4539-a484-ad138594c3e0">

Reply:
Yeah there's conflicting information [here](https://docs.yubico.com/hardware/yubikey/yk-tech-manual/yk5-apps.html#piv-aes-management-key), which is where I found 5.4.2. I think you'd have to find someone with a 5.4.0 or 5.4.1 key to confirm this.

Reply:
That's entertaining. That article links to the page I linked "for more details". My suspicion is that they don't actually sell keys with every point release of firmware. They may have added it in 5.4.2, but maybe 5.4.3 was the first 5.4 release to make it into keys they sold. This would effectively make both docs correct in saying all 5.4 keys support it, and that it was added in 5.4.2.

I may reach out to Yubico support to try to get clarification.

Reply:
It appears that 5.4.2 is in fact the first 5.4 release to be sold based on the following 2 articles. Ten days _after_ announcing 5.4.2 as the initial release for yubikey 5 fips they announce the shipping of 5.4 keys.

5.4.2 is the initial firmware version for yubikey 5 fips keys (article 4 May, 2021)
https://support.yubico.com/hc/en-us/articles/360021443340-YubiKey-5-NFC-FIPS

5.4 started shipping on yubikey 5 series keys (article 14 May, 2021)
https://www.yubico.com/blog/yubikey-firmware-update-yubikey-5-series-with-firmware-5-4/

I think this version check is good to leave as-is.